pkgs/tools/misc/nbqa/default.nix

```diff
@@ -0,0 +1,93 @@
+{ black
+, blacken-docs
+, fetchFromGitHub
+, lib
+, python3
+, ruff
+}:
+python3.pkgs.buildPythonApplication rec {
+  pname = "nbqa";
+  version = "1.7.0";
+
+  src = fetchFromGitHub {
+    owner = "nbQA-dev";
+    repo = "nbQA";
+    rev = version;
+    hash = "sha256-CTF5HisBS44Ta18cVT04UrMIF30WmEBwZBGW7fkKXwk=";
+  };
+
+  passthru.optional-dependencies = {
+    # Based on https://github.com/nbQA-dev/nbQA/blob/1.7.0/setup.cfg#L47-L56
+    black = [ black ];
+    blacken-docs = [ blacken-docs ];
+    flake8 = [ python3.pkgs.flake8 ];
+    isort = [ python3.pkgs.isort ];
+    jupytext = [ python3.pkgs.jupytext ];
+    mypy = [ python3.pkgs.mypy ];
+    pylint = [ python3.pkgs.pylint ];
+    pyupgrade = [ python3.pkgs.pyupgrade ];
+    ruff = [ ruff ];
+  };
+
+  propagatedBuildInputs =
+    [
+      # Based on https://github.com/nbQA-dev/nbQA/blob/1.7.0/setup.cfg#L31-L35
+      python3.pkgs.autopep8
```

Comment:
`with python3.pkgs;`

Reply:
I prefer to not use `with`, since it [screws with scope](https://nix.dev/recipes/best-practices#with-scopes).

Reply:
greetings ! the link you provided doesn't seem to apply in this case.

there is also just statistically ( for better or worse ) more examples of `with pkgs etc` than `[ pkgs.a pkgs.b ]`

Reply:
`git grep 'with pkgs' | wc -l` returns 1276 results in nixpkgs, which I suspect is very much less than the majority of the 80,000+ packages in there.

Reply:
yes, that is because `nixpkgs` mainly uses `callPackage` pattern and usually dependencies are passed as arguments.

however in many cases such for python dependencies, often you have to get them from `pythonPackages`

`rg 'with python'`
 

use of `with` is discouraged only on top level ( eg `with pkgs; mkDerivation` ) when used in examples such as 
`
nativeBuildInputs  = with pkgs ; []
`

with only modifies scope of that list

Reply:
I'd like to add that we rarely see the pattern without `with` in the **well-maintained** packages that we usually review.

Reply:
Cool, I'll use `with` since you both recommend it.